database/src/main/java/com/firebase/ui/database/ObservableSnapshotArray.java

```diff
@@ -24,4 +25,11 @@
     public ObservableSnapshotArray(@NonNull SnapshotParser<T> parser) {
         super(new CachingSnapshotParser<>(parser));
     }
+
+    /**
+     * Method for changing/updating the {@link Query} in existing adapter.
+     *
+     * @param newQuery is a new updated query.
+     */
+    public abstract void updateQuery(@NonNull Query newQuery);
```

Comment:
Should this be added in `BaseObservableSnapshotArray` instead? It seems a bit strange that this would be the only method defined on `ObservableSnapshotArray`

Reply:
Because `BaseObservableSnapshotArray` is inherited in both `ObservableSnapshotArray` of Firestore as well as Firebase realtime database too. And most importantly, the `Query` class of both databases are not the same.
- Firestore Query Class: `com.google.firebase.firestore.Query`
- Firebase Query Class: `com.google.firebase.database.Query`

Reply:
If we add this method in `BaseObservableSnapshotArray` then `Query` parameter will be different for them. What can we do?

Reply:
I would say let's just not make this part of the base class at all, it can just be something implemented in each of the `Array` classes.

That will also allow us to give the indexed version a better name: `updateKeyQuery()`

Reply:
We have instantiated `FirestoreArray` by its superclass references in `Adapter` classes. So, Will it be fine them to be instantiated as `Array` classes?

Reply:
Ah ok I see the issue.  Let me think about it for a minute (I'll clone the branch and play around)

Reply:
Ok yeah I think that's the best solution, instantiate it as an `Array` but have the `getSnapshots()` method still return `ObservableSnapshotArray` so that people only call `updateQuery` on the `*Adapter` classes.

Reply:
Fine 😃. Okay then 👍 .

Reply:
I have a question. How could we know that when to call `updateQuery()` or `updateKeyQuery()`?
In `FirebaseRecyclerAdapter` if I create an instance of `FirebaseArray` then we can't instantiate `FirebaseIndexArray`. How to do it?

Reply:
Oh no that is a problem ... any ideas?  I don't know why I am so opposed to adding `updateQuery` to the `ObservabeSnapshotArray` classes ... it just feels wrong.  But maybe that is the best way.

Reply:
I also thought a lot about it before implementation in `ObservableSnapshotArray` class. But I can't find anything helpful.

Reply:
I had one idea. But that idea was not much good to implement. Means, it's not good practice to implement. I had idea to use `instanceof` to check whether it's instance of Firebase array or index array. We will check it in our method and by casting we will call array class method. That was the idea.